rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes tables in memory which which have not been flushed and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables. The latest records will have precedence following
+the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. Additionally, `MemTableIterator` and 
+`SstIterator` provide ready-to-use capabilities to scan tables which may have keys
+within the range. 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the query began scanning its range of records. For example,
+a query may include only a subset of records that were inserted in the same 
+`WriteBatch`. In other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-????](https://github.com/slatedb/slatedb/pull/278) in order to prevent the cleanup
```

Comment:
Would the manifest with the checkpoint reference be persisted? There was discussion about this in discord. One option is to just keep it in memory, because range queries could be often and `min_age` property of GC will guard against deletions to a good extent.

Reply:
The proposal in #278 is to maintain a checkpoint as part of the regular manifest polling that the reader/writer do, and IIUC the proposal here is to use this checkpoint for range scans  - so yes the checkpoint would be persisted in the manifest. There shouldn't be much added cost since the checkpoint is shared and written to the manifest at the poll interval or as part of regular manifest updates (e.g. for adding new L0 ssts).